### PR TITLE
feat: alt way for auth media (BREAKING)

### PR DESCRIPTION
**Pull request by td-famedly**
This removes the version checks, and expects clients to manually return both endpoints using the `useAuthentication` flag

Breaking change but tries both new and old endpoints. Would require clients to implement a retry mech for their media if they fail.

Particularly useful when servers are misconfigured so they return on the old endpoints but not on the new ones. (not a fan, just configure them correctly!)

### CODE REVIEW

File: lib/src/utils/uri_extension.dart [hunk 1]
```diff
@@ -59,19 +62,22 @@ extension MxcUriExtension on Uri {
   ///
   /// Important! To use this link you have to set a http header like this:
   /// `headers: {"authorization": "Bearer ${client.accessToken}"}`
-  Future<Uri> getThumbnailUri(Client client,
-      {num? width,
-      num? height,
-      ThumbnailMethod? method = ThumbnailMethod.crop,
-      bool? animated = false}) async {
+  Uri getThumbnailUri(
```
Comment (nico-famedly) [inline lib/src/utils/uri_extension.dart:65]: I think we should deprecate the functions to get a raw URL as an SDK user. They might be useful internally, but SDK users will need to manually set auth headers when using any of these URIs. This means you basically want to use the thumbnail and download functions from the SDK anyway, never really the *Uri functions, I think.

Comment (nico-famedly) [reply lib/src/utils/uri_extension.dart:65]: Okay, so some clients are using this to feed Image.network and similar, so those would benefit from keeping these around. However I think we can at least deprecate the extensions on Uri and only expose 2 functions returning a Uri instead if 4+.

**Review by nico-famedly — APPROVED** [review]

File: lib/src/client.dart [hunk 2]
```diff
@@ -1189,21 +1178,25 @@ class Client extends MatrixApi {
...
+        unexpectedResponse(response, responseBody);
+      }
+      final responseString = utf8.decode(responseBody);
+      final json = jsonDecode(responseString);
+      return ServerConfig.fromJson(json as Map<String, Object?>);
+    }
+
+    try {
+      return request('_matrix/client/v1/media/config');
+    } catch (e) {
```
Comment (krille-chan) [inline lib/src/client.dart:1197]: Do we know the exact error message that would be returned here? Then we should listen to only this and rethrow other errors

File: lib/src/client.dart [hunk 3]
```diff
@@ -1222,27 +1215,30 @@ class Client extends MatrixApi {
...
+        unexpectedResponse(response, responseBody);
+      }
+      return FileResponse(
+          contentType: response.headers['content-type'], data: responseBody);
+    }
+
+    try {
+      return request(
+          '_matrix/client/v1/media/download/${Uri.encodeComponent(serverName)}/${Uri.encodeComponent(mediaId)}');
+    } catch (e) {
```
Comment (krille-chan) [inline lib/src/client.dart:1238]: Same here

File: lib/src/client.dart [hunk 4]
```diff
@@ -1266,26 +1262,30 @@ class Client extends MatrixApi {
...
+        unexpectedResponse(response, responseBody);
+      }
+      return FileResponse(
+          contentType: response.headers['content-type'], data: responseBody);
+    }
+
+    try {
+      return request(
+          '_matrix/client/v1/media/download/${Uri.encodeComponent(serverName)}/${Uri.encodeComponent(mediaId)}/${Uri.encodeComponent(fileName)}');
+    } catch (e) {
```
Comment (krille-chan) [inline lib/src/client.dart:1285]: And here

**Review by krille-chan — CHANGES_REQUESTED** [review]
Don't get it yet... why should we implement such a hacky workaround because of a misconfigured server? Shouldn't we fix the server instead?

**Pull request closed without merging**